forem/forem
# djuber/ruby3

PR is to merge into @citizen428's PR branch so I don't know how much of this I'm going to fill out.


## What type of PR is this? (check all applicable)


## Description

I don't understand it fully yet - but this seems to resolve (temporarily) the issue where rails app `initialize!` call hangs (stack traces suggest the issue was occuring when calling either `protected_instance_methods` or `public_instance_methods` on a Hash delegates superclass, but I didn't get much further in the search for answers.) Almost every specific idea I have suggests that Hashie is the root cause or closely related to the cause - (it's relied on by a few gems, including buffer and omniauth, and removing omniauth also made this go away when I was testing) - I've seen the location of the infinite loop happen alternately in `ActionDispatch::Response::Headers < DelegateClass(Hash)` or AmazingPrint or elsewhere. 

I repeated @maestromac bisection search over the dependencies to zero in on omniauth-oauth as a problem, but I'm still confused what the actual cause is - all stack traces did point to the tsort/each_strongly_connected_component being related to the behavior described in https://github.com/forem/forem/pull/12103/#issuecomment-806388905 - I think it's interesting that mac and I found a _different_ problem gem that removing made the situation better. That might be a sensitivity to bisection (depending on where you make the cuts, you'll find a different keystone, which further supports the suspicion we have some cyclic dependency somehow?)


- Reinstall buffer (doesn't seem to have been the culprit, unless it was only _added_ as a test solution to resolve the issue).
- Update all gems that differed from a clean installation (new vanilla rails project, copy Gemfile from forem ruby3 branch, run bundle) Gemfile.lock and the forem Gemfile.lock

## Related Tickets & Documents

https://github.com/forem/forem/pull/12103/

## QA Instructions, Screenshots, Recordings

Pull branch
`bin/setup` should complete successfully (install bundle dependencies, yarn, and prepare db and es)
Problem observed previously is rails initialization entered 100% cpu loop.

If I remove buffer the issue definitely comes back (setup hangs initializing the app).

My smallest reproduction step (bypassing all of the binstubs and just initializing the app itself) has been:

```ruby
require File.expand_path("config/application", __dir__)
@app = PracticalDeveloper::Application.new
@app.initialize! 
```

However, that just gets us back to "booting rails hangs"

more specifically, I'm always finding the sticky point (`rbspy snapshot -p PID` is useful here):

Forem with buffer gem removed from this branch:

```
require [c function] - (unknown):0
<main> - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/rspec-core-3.10.1/lib/rspec/core/metadata.rb:1
<module:RSpec> - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/rspec-core-3.10.1/lib/rspec/core/metadata.rb:498
<module:Core> - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/rspec-core-3.10.1/lib/rspec/core/metadata.rb:497
<module:HashImitatable> - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/rspec-core-3.10.1/lib/rspec/core/metadata.rb:445
block (2 levels) in <class:Class> - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/amazing_print-1.3.0/lib/amazing_print/core_ext/class.rb:23
call [c function] - (unknown):0
public_instance_methods [c function] - (unknown):0
```

Vanilla rails with this Gemfile (without buffer):

```
require [c function] - (unknown):0
<main> - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/actionpack-6.1.3.1/lib/action_controller/metal.rb:8
require - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/activesupport-6.1.3.1/lib/active_support/dependencies.rb:334
load_dependency - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/activesupport-6.1.3.1/lib/active_support/dependencies.rb:304
block in require - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/activesupport-6.1.3.1/lib/active_support/dependencies.rb:332
require - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/zeitwerk-2.4.2/lib/zeitwerk/kernel.rb:43
require - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/bootsnap-1.7.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:46
require_with_bootsnap_lfi - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/bootsnap-1.7.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:25
register - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/bootsnap-1.7.3/lib/bootsnap/load_path_cache/loaded_features_index.rb:114
block in require_with_bootsnap_lfi - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/bootsnap-1.7.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:24
require [c function] - (unknown):0
<main> - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/actionpack-6.1.3.1/lib/action_dispatch/http/response.rb:8
<module:ActionDispatch> - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/actionpack-6.1.3.1/lib/action_dispatch/http/response.rb:540
<class:Response> - /home/djuber/src/forem/vendor/cache/ruby/3.0.0/gems/actionpack-6.1.3.1/lib/action_dispatch/http/response.rb:537
DelegateClass - /home/djuber/.rbenv/versions/3.0.0/lib/ruby/3.0.0/delegate.rb:444
protected_instance_methods [c function] - (unknown):0
```
